matchtree.go

```diff
@@ -1375,3 +1381,26 @@ func pruneMatchTree(mt matchTree) (matchTree, error) {
 	}
 	return mt, err
 }
+
+// isRegexpAll returns true if the query matches all possible lines.
+//
+// Note: it is possible for a funky regex to actually match all but this
+// returns false. This returns true for normal looking regexes like ".*" or
+// "(?-s:.*)".
+func isRegexpAll(r *syntax.Regexp) bool {
+	// Note: we don't care about any flags since we are looking for .* and we
+	// don't mind if . matches all or everything but newline.
+
+	// Our main target: .*
+	if r.Op == syntax.OpStar && len(r.Sub) == 1 { // *
+		// (?s:.) or (?-s:.)
+		return r.Sub[0].Op == syntax.OpAnyChar || r.Sub[0].Op == syntax.OpAnyCharNotNL
+	}
+
+	// Strip away expressions being wrapped in paranthesis
+	if (r.Op == syntax.OpCapture || r.Op == syntax.OpConcat) && len(r.Sub) == 1 {
```

Comment:
Maybe we could convert this to a loop as it's tail recursive. (This is likely fine, but recursion on flexible user input always makes me slightly nervous!)

Reply:
I actually originally wrote this iteratively, but thought this was more obvious. However, I liked the iterative approach so will update it back to that implementation :)